### PR TITLE
Consume

### PR DESCRIPTION
**Pull request by SimonRichardson**
Consume allows the consumption of a offer. The following PR requires the following
juju PRs to land. In order to allow consumption of the offer the following changes had
to be done:

 1. Expose all the properties of `FullStatus`, this was originally done for backwards
compatibility, but is now hurting us in the ability to move forward with new features.
 1. To prevent work arounds where anonymous embedded structs in api server params
of juju go code would not correctly reveal the arguments, the following PRs where 
required:
     1. https://github.com/juju/juju/pull/10471
     1. https://github.com/juju/juju/pull/10467
     1. https://github.com/bcsaller/jsonschema/pull/1

The last PR for jsonschema seems to be abandoned and work to move to the upstream
repo is required. I've forked jsonschema for now, until I understand the changes required
to move upstream. It maybe we don't and just fork the current jsonschema in to the 
juju/juju namespace. The issue we have there is that there already exists a juju/jsonschema,
so we'll have to rename the package.

The previous code would push the embedded results from the response into the
`kwargs` (`unknow_fields`), which made it trivial to get embedded results. Unfortunately
sending new arguments would mean we would have to curate the arguments by hand,
this has two frailties:
 1. We loose the strong typing that we get from the definitions (why have them if we have
to hack around them all the time).
 1. If any changes up stream to the definitions/parameters means that subtle bugs can
be introduced because we've overridden the definitions and created our own.

Note: the large number of changes are because we've had to regenerate the schemas
because of the new exposed arguments from embedding. 

## QA

```
juju bootstrap lxd test --no-gui
tox -e example -- examples/crossmodel.py
tox -e integration -- tests/integration/test_crossmodel.py
```

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/WU3cCfcs/1668-cmr-pylibjuju-consume-application

File: juju/model.py [hunk 1]
```diff
@@ -1912,6 +1918,111 @@ def charmstore(self):
...
+        if url.source == "":
+            current = await self.get_controller()
+            url.source = current.controller_name
+        await controller.connect(controller_name=url.source)
+        return controller
+
+
+def _create_consume_args(offer, macaroon, controller_info):
+    """
+    Convert a typed object that has been normalised to a overrided typed
```
Comment (mitechie) [inline juju/model.py]: can you provide an example thing here to get a visual idea of the transforming expected please?

File: examples/crossmodel.py [hunk 2]
```diff
@@ -8,45 +8,74 @@
...
-        await model.create_offer("mysql:db")
+        await model_1.create_offer("mysql:db")
 
-        offers = await model.list_offers()
-        await model.block_until(
+        offers = await model_1.list_offers()
+        await model_1.block_until(
             lambda: all(offer.application_name == 'mysql'
-                        for offer in offers))
+                        for offer in offers.results))
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply examples/crossmodel.py]: Unfortunately not, this is from the generated code.

File: examples/crossmodel.py [hunk 3]
```diff
@@ -8,45 +8,74 @@
...
+
+        print('Show offers', ', '.join("%s: %s" % item for offer in offers.results for item in vars(offer).items()))
+
+        print('Consuming offer')
+        await model_2.consume("admin/test-cmr-1.mysql")
 
-        print('Show offers', ', '.join("%s: %s" % item for offer in offers for item in vars(offer).items()))
+        status = await model_2.get_status()
+        if 'mysql' not in status.remote_applications:
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply examples/crossmodel.py]: Nope, it was help testing, I'll remove

File: juju/client/_client.py [hunk 4]
```diff
@@ -66,6 +66,20 @@ def from_connection(cls, connection):
...
+        """
+        Returns the best facade version for a given facade. This will help with
+        trying to provide different functionality for different facade versions.
+
+        @param connection: initialized Connection object.
+        """
+        facade_name = cls.__name__
+        if not facade_name.endswith('Facade'):
+           raise TypeError('Unexpected class name: {}'.format(facade_name))
+        facade_name = facade_name[:-len('Facade')]
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/client/_client.py]: Try, it's just symmetrical with the above class method

File: juju/model.py (hunk 1 above)
Comment (SimonRichardson) [reply juju/model.py]: So the issue here is that macaroons aren't exposed correctly, so we have to fill in the blanks

File: juju/model.py [hunk 5]
```diff
@@ -1912,6 +1918,111 @@ def charmstore(self):
...
+        """
+        Adds a remote offer to the model. Relations can be created later using
+        "juju relate".
+        """
+        try:
+            offer = parse_offer_url(endpoint)
+        except OfferParseError as e:
+            log.error(e.message)
+            raise
+        if offer.has_endpoint():
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/model.py]: So does the offer still include any remaining endpoints, if it does then it's an invalid.

**Comment by SimonRichardson** [comment]
@mitechie we'll need the other PRs landing before this will be able to land.

File: examples/crossmodel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mitechie) [reply examples/crossmodel.py]: ah ok, that sucks

File: juju/client/_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mitechie) [reply juju/client/_client.py]: Yea, I figured I'd not ping it in two places but gotcha

File: juju/model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mitechie) [reply juju/model.py]: ok, ty

**Comment by mitechie** [comment]
> @mitechie we'll need the other PRs landing before this will be able to land.

Understand

File: examples/crossmodel.py [hunk 6]
```diff
@@ -8,45 +8,72 @@
...
+        application = await offering_model.deploy(
             'mysql',
             application_name='mysql',
             series='trusty',
             channel='stable',
         )
 
         print('Waiting for active')
-        await model.block_until(
+        await offering_model.block_until(
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply examples/crossmodel.py]: TBH you don't have to wait and should work, it just following the patterns in the examples.

File: examples/crossmodel.py [hunk 7]
```diff
@@ -8,45 +8,72 @@
...
+
+        print('Show offers', ', '.join("%s: %s" % item for offer in offers.results for item in vars(offer).items()))
+
+        print('Consuming offer')
+        await consuming_model.consume("admin/test-cmr-1.mysql")
 
-        print('Show offers', ', '.join("%s: %s" % item for offer in offers for item in vars(offer).items()))
+        status = await consuming_model.get_status()
+
+        time.sleep(10)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply examples/crossmodel.py]: Sure, can update that.

File: examples/crossmodel.py [hunk 8]
```diff
@@ -8,45 +8,72 @@
...
+
+    except Exception:
+        log.exception("Example failed!")
+        raise
+
     finally:
-        print('Disconnecting from model')
-        await model.disconnect()
+        print('Disconnecting from controller')
+        await controller.disconnect()
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply examples/crossmodel.py]: So pylib can't bootstrap, I was surprised about this if I'm honest. So what it does is it latches on to any controller that's available using the credentials/controller yaml found locally. So we don't ever bootstrap or destroy anything as we don't own it.

File: juju/client/_client.py [hunk 9]
```diff
@@ -66,6 +66,20 @@ def from_connection(cls, connection):
 
         return c
 
+    @classmethod
+    def best_facade_version(cls, connection):
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/client/_client.py]: This is using the same terminology that juju uses, for familiarity

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**